cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -286,6 +312,135 @@ private void deployApp(SystemApplication application) throws Exception {
                  application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
   }
 
+  @VisibleForTesting
+  void autoInstallResources(String capability, List<AutoInstall> autoInstall) {
+    if (autoInstall == null || autoInstall.isEmpty()) {
+      LOG.debug("Capability {} do not have auto install resources associated with it", capability);
+      return;
+    }
+
+    for (AutoInstall a : autoInstall) {
+      Package[] packages;
+      String hubUrl = a.getHubUrl();
+      try {
+        URL url = new URL(new URL(hubUrl), "/v2/packages.json");
+        String packagesJson = doGetAsString(url);
+        // Deserialize packages.json from hub
+        // See https://cdap.atlassian.net/wiki/spaces/DOCS/pages/554401840/Hub+API?src=search#Get-Hub-Catalog
+        packages = GSON.fromJson(packagesJson, Package[].class);
+      } catch (Exception e) {
+        LOG.error("Failed to get packages.json from {}", hubUrl, e);
+        return;
+      }
+
+      String currentVersion = getCurrentVersion();
+      // Get the latest compatible version of each plugin from hub and install it
+      Arrays.stream(packages)
+        .filter(p -> p.versionIsInRange(currentVersion))
+        .collect(Collectors.groupingBy(Package::getName,
+                                       Collectors.maxBy(Comparator.comparing(Package::getArtifactVersion))))
+        .values()
+        .stream()
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .forEach(p -> {
+          try {
+            installPlugin(new URL(hubUrl), p);
+          } catch (Exception e) {
+            LOG.error("Failed to install plugin {}", p.getName(), e);
+          }
+        });
+    }
+  }
+
+  @VisibleForTesting
+  String getCurrentVersion() {
+    return VersionHelper.getCDAPVersion().getVersion();
+  }
+
+  private void installPlugin(URL url, Package pkg) throws IOException {
+    // Deserialize spec.json
+    URL specUrl = new URL(url, "/v2/packages/" + pkg.getName() + "/"
```

Comment:
This logic should move in the Package class.

Reply:
fixed